File: LR3/RecursiveDescent/src/grammar/Grammar.cpp
#include "../../include/grammar/grammar.h"


Grammar::Grammar(
	const symbol_vector& nonTerms, 
	const symbol_vector& terms, 
	const symbol_ptr& start, 
	const rule_vector& prods) : Grammar()
{
	if (!start)
	{
		throw std::invalid_argument{ "Null axiom." };
	}

	if (std::any_of(nonTerms.begin(), nonTerms.end(), [&](auto s) { return !s; }))
	{
		throw std::invalid_argument{ "Null symbol in nonterminals." };
	}

	if (std::any_of(terms.begin(), terms.end(), [&](auto s) { return !s; }))
	{
		throw std::invalid_argument{ "Null symbol in terminals." };
	}

	if (std::any_of(prods.begin(), prods.end(), [&](auto p) { return !p; }))
	{
		throw std::invalid_argument{ "Null rule in prods." };
	}

	nonTerminals = nonTerms;
	terminals = terms;
	axiom = start;
	rules = prods;
}

const symbol_ptr& Grammar::getSymbol(const std::string& name) const
{
	for (auto& s : terminals) { if (s->getName() == name) return s; }
	for (auto& s : nonTerminals) { if (s->getName() == name) return s; }

	throw std::invalid_argument{ "No symbol with specified name." };
}

File: LR3/recursive-descend/src/main/scala/grammar/Grammar.scala
package grammar

// Grammar symbol type
object SymbolType extends Enumeration {
    type SymbolType = Value
    val Term, NonTerm, Eps = Value
}

// Grammar symbol
class Symbol(val name: String, val spell: Option[String], val stype: SymbolType.SymbolType = SymbolType.Term) {
    if (name == null || name.isBlank()) throw new Exception("Invalid name")
    if (stype == null) throw new Exception("Invalid symbol type")
}

class Rule(val lhs: List[Symbol], val rhs: List[Symbol]) {
    if (lhs == null || lhs.isEmpty) throw new Exception("Invalid left-hand side of the rule")
    if (rhs == null) throw new Exception("Invalid right-hand side of the rule")
}

class Grammar(val terms: List[Symbol], val nonTerms: List[Symbol], val rules: List[Rule], val axiom: Symbol) {
    if (terms == null || terms.isEmpty) throw new Exception("Invalid terms list")
    if (nonTerms == null || nonTerms.isEmpty) throw new Exception("Invalid nonterms list")
    if (rules == null || rules.isEmpty) throw new Exception("Invalid rules list")
    if (axiom == null) throw new Exception("Invalid axiom")
}